ykrt/src/thread_intercept.rs

```diff
@@ -1,11 +1,54 @@
 use std::os::raw::{c_int, c_void};
 
+use std::{ffi::CString, ptr::null_mut};
+
+use libc::{dlsym, free, malloc, pthread_attr_t, pthread_create, pthread_t};
+
+const SHADOW_STACK_SIZE: usize = 1000000;
+
+#[derive(Debug)]
+struct ThreadRoutine {
+    pub args: *mut c_void,
+    pub start_routine: extern "C" fn(*mut c_void) -> *mut c_void,
+}
+
+extern "C" fn wrap_thread_routine(arg: *mut c_void) -> *mut c_void {
+    let str = CString::new("shadowstack_0").unwrap();
+    unsafe {
+        //  Obtain address of a shadowstack_0 symbol
+        let shadowstack_addr: *mut libc::c_void = dlsym(null_mut(), str.as_ptr() as *const i8);
+        if shadowstack_addr.is_null() {
+            panic!("Unable to find shadowstack address!")
+        }
+        // Allocate stack
+        let stack_addr = malloc(SHADOW_STACK_SIZE);
+        if stack_addr.is_null() {
+            panic!("Unable allocate stack!")
+        }
+        // Set shadowstack addr with new allocated stack
+        *(shadowstack_addr as *mut *mut c_void) = stack_addr;
+        let thread_routine: &ThreadRoutine = (arg as *mut ThreadRoutine)
+            .as_ref()
+            .expect("Thread routine function and args data.");
+        // Call original thread routine
+        let result = (thread_routine.start_routine)(thread_routine.args);
+        free(stack_addr);
```

Comment:
Is there a benefit of using POSIX functions for thread cleanup such as `pthread_cleanup_push` and `pthread_cleanup_pop`?
These functions are not part of rust [libc crate](https://docs.rs/libc/latest/libc/)

Reply:
Can we solve this issue by calling the "normal" `pthread_create` function? If not, then I guess we have to emulate whatever `pthread_create` does... At least we would probably have to add a wrapper around `pthread_cleanup_push` that panics, so that if any code does use it, it will die in a clear and obvious way.

Reply:
I am not sure I understand what you mean by "calling the normal pthread_create function"
I thought that's what I'm doing here - calling POSIX `pthread_create` function with a wrapper function that will call the original `pthread_create` `start_routine`

Reply:
I don't think I understood this fully. What I should have said is: let's add a wrapper around `pthread_cleanup_push` that panics if anyone uses it. That way we'll at least know if we ever need to support it! [I don't mind hard errors which make it clear where we need to look for the solution, but I am always very scared if errors can get through silently!]

Reply:
I think I understand what you mean, I can definetly add a wrapper for `pthread_cleanup_push` and `pthread_cleanup_pop` functions.

But my question was -  Should we use these `pthread_cleanup_push/pop` functions for YK-created threads cleanup? In this PR context, the the cleanup is `free(stack_addr)`.
I guess there can be an case where the thread is cancelled and we end up not freeing the allocated stack.

Reply:
> But my question was - Should we use these pthread_cleanup_push/pop functions for YK-created threads cleanup?

Good question! I don't think we need to since our wrapper automatically means we can do the "last" cleanup ourselves?

Reply:
We decided to not handle this edge case of thread cancelation although we understand that it might be a potential issue the impact is not clear. I added `YKFIXME`  as a reminder.